File: src/core_ffi/rust_ffi/adapter/short_code.cc
#include "rust_ffi/include/short_code.h"

#include <short_code/short_code.h>

using klotski::codec::ShortCode;

using klotski::ffi::RsShortCode;
using klotski::ffi::RsCommonCode;

bool klotski::ffi::short_code_check(const uint32_t val) {
    return ShortCode::check(val);
}

uint32_t klotski::ffi::short_code_from_str(const rust::Str s) {
    const std::string_view sv {s.data(), s.length()};
    if (const auto ret = ShortCode::from_string(sv); ret.has_value()) {
        return ret.value().unwrap();
    }
    return 0xFFFFFFFF; // return invalid value for now
}

rust::String RsShortCode::to_string() const noexcept {
    return ShortCode::unsafe_create(code).to_string();
}

RsCommonCode RsShortCode::to_common_code() const noexcept {
    return {ShortCode::unsafe_create(code).to_common_code().unwrap()};
}

void klotski::ffi::short_code_speed_up(const bool fast_mode) {
    ShortCode::speed_up(fast_mode);
}

File: src/core_ffi/rust_ffi/src/short_code.rs
use crate::ffi;

#[derive(Debug)]
pub struct ShortCode {
    code: u32
}

impl ShortCode {
    // fn new(code: u32) -> Option<ShortCode> {
    //     if !Self::check(code) {
    //         return None
    //     }
    //     Some(Self::unsafe_new(code))
    // }

    pub fn unsafe_new(code: u32) -> ShortCode {
        Self {code}
    }

    fn check(code: u32) -> bool {
        ffi::short_code_check(code)
    }
}
